### PR TITLE
CCAP-138: Added OneDrive transfer support.

### PR DESCRIPTION
**Pull request by jamesiarmes**
#### 🔗 Jira ticket

CCAP-138

#### ✍️ Description

Added `/transfer` endpoint with support for a url source and OneDrive destination.

#### 📷 Design reference

[Tech Spec](https://www.notion.so/cfa/Tech-Spec-Document-Transfer-Microservice-d89aa25d41e447b3920ff07ab03ce7c1?pvs=4)

#### 🧪 Testing instructions

- [ ] Checkout the code and run `bundle install` to install dependencies
- [ ] Set the `ONEDRIVE_CLIENT_ID`, `ONEDRIVE_CLIENT_SECRET`, `ONEDRIVE_TENANT_ID`, and `ONEDRIVE_DRIVE_ID` [environment variables](https://github.com/codeforamerica/document-transfer-service/blob/ccap-138-onedrive/doc/destination.md#configuration) using the values in LastPass 
- [ ] Run `bundle exec rackup` (or `docker compose up`) to start the API
- [ ] Run the following to transfer a document to onedrive

   ```bash
   curl -H 'Content-Type: application/json' \
   -d '{ "source": { "type": "url", "url": "https://file-examples.com/storage/fe83e1f11c664c2259506f1/2017/10/file-example_PDF_500_kB.pdf" }, "destination": { "type": "onedrive", "path": "pr/review", "filename": "review-test.pdf" } }' \
   -X POST \
   http://localhost:9292/transfer
   ```

#### ✅ Completion tasks

- [x] Added relevant tests
- [x] Meets acceptance criteria

### CODE REVIEW

File: doc/destination.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+```json
+{
+  "status": "success",
+  "destination": "onedrive",
+  "path": "/document/path/document.pdf"
+}
+```
+
+[onedrive]: https://www.microsoft.com/en-us/microsoft-365/onedrive/onedrive-for-business
```
Comment (ShaynaCummings) [inline doc/destination.md]: Could also add a link to [this](https://learn.microsoft.com/en-us/graph/api/driveitem-put-content?view=graph-rest-1.0&tabs=http), found it helpful when I was first learning about the onedrive api

File: doc/destination.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+With the OneDrive destination, the service will upload the document to a
+configured [Microsoft OneDrive][onedrive] account.
+
+### Configuration
+
+_**The OneDrive destination requires additional configuration of the service in
+order to function properly.**_
+
+Before you can use the OneDrive destination, you must create a new application
```
Comment (ShaynaCummings) [inline doc/destination.md]: I would phrase this as "must create a new application (Azure Portal -> App Registrations) to get OneDrive API credentials"

File: doc/destination.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+authenticating.
+
+The following environment variables must be set on the service:
+
+| Name                   | Description                                                         |
+|------------------------|---------------------------------------------------------------------|
+| ONEDRIVE_CLIENT_ID     | The client ID of the OneDrive application.                          |
+| ONEDRIVE_CLIENT_SECRET | The client secret of the OneDrive application.                      |
+| ONEDRIVE_DRIVE_ID      | The drive ID of the OneDrive account documents will be uploaded to. |
+| ONEDRIVE_TENANT_ID     | The tenant ID of the OneDrive application.                          |
```
Comment (ShaynaCummings) [inline doc/destination.md]: Suggestion: change description to `The directory (tenant) ID of the OneDrive application` to match the app registration nomenclature

File: doc/destination.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+Before you can use the OneDrive destination, you must create a new application
+in the Azure Portal, with read and write permissions to the appropriate OneDrive
+account. Created a client id and secret for the application to use for
+authenticating.
+
+The following environment variables must be set on the service:
+
+| Name                   | Description                                                         |
+|------------------------|---------------------------------------------------------------------|
+| ONEDRIVE_CLIENT_ID     | The client ID of the OneDrive application.                          |
```
Comment (ShaynaCummings) [inline doc/destination.md]: Suggestion: change description to `The application (client) ID of the OneDrive application` to match the app registration nomenclature

File: doc/destination.md (hunk 3 above)
Comment (jamesiarmes) [reply doc/destination.md]: Tenant and client are the appropriate nomenclature and how these IDs are referred to in most of the documentation for both the API and the authentication library.

**Comment by cy-by** [comment]
@jamesiarmes rec:

new file called `sample.env` with:

```
export ONEDRIVE_CLIENT_ID=
export ONEDRIVE_CLIENT_SECRET=
export ONEDRIVE_TENANT_ID=
export ONEDRIVE_DRIVE_ID=

```

**Review by cy-by — APPROVED** [review]
Paired on my concerns with James, approved!

**Review by cy-by — APPROVED** [review]
Approved

**Pull request merged**